node/src/service.rs

```diff
@@ -195,6 +199,161 @@ pub fn new_partial(
+    let sync_keystore = keystore_container.sync_keystore();
+    let initial_assigned_para_id = Arc::new(Mutex::new(None));
+
+    let telemetry_worker_handle = telemetry.as_ref().map(|(worker, _)| worker.handle());
+
+    let telemetry = telemetry.map(|(worker, telemetry)| {
+        task_manager
+            .spawn_handle()
+            .spawn("telemetry", None, worker.run());
+        telemetry
+    });
+
+    let transaction_pool = sc_transaction_pool::BasicPool::new_full(
+        config.transaction_pool.clone(),
+        config.role.is_authority().into(),
+        config.prometheus_registry(),
+        task_manager.spawn_essential_handle(),
+        client.clone(),
+    );
+
+    let block_import = ParachainBlockImport::new(client.clone(), backend.clone());
+    // The nimbus import queue ONLY checks the signature correctness
+    // Any other checks corresponding to the author-correctness should be done
+    // in the runtime
+    let import_queue = nimbus_consensus::import_queue(
+        client.clone(),
+        block_import.clone(),
+        move |block_hash, ()| {
+            let client_set_aside_for_cidp = client_set_aside_for_cidp.clone();
+            let sync_keystore = sync_keystore.clone();
+            let initial_assigned_para_id = initial_assigned_para_id.clone();
+            let cc_spawn_tx = cc_spawn_tx.clone();
+
+            async move {
+                check_assigned_para_id(
+                    cc_spawn_tx,
+                    &*sync_keystore,
+                    initial_assigned_para_id,
+                    client_set_aside_for_cidp,
+                    block_hash,
+                )?;
+
+                let time = sp_timestamp::InherentDataProvider::from_system_time();
+
+                Ok((time,))
+            }
+        },
+        &task_manager.spawn_essential_handle(),
+        config.prometheus_registry(),
+        false,
+    )?;
+
+    let maybe_select_chain = None;
+
+    Ok(PartialComponents {
+        backend,
+        client,
+        import_queue,
+        keystore_container,
+        task_manager,
+        transaction_pool,
+        select_chain: maybe_select_chain,
+        other: (block_import, telemetry, telemetry_worker_handle),
+    })
+}
+
+/// Check the parachain assignment using the orchestrator chain client, and send a `CcSpawnMsg` if
+/// the para id has changed since the last call to this function.
+fn check_assigned_para_id<Client>(
+    cc_spawn_tx: UnboundedSender<CcSpawnMsg>,
+    sync_keystore: &dyn SyncCryptoStore,
+    initial_assigned_para_id: Arc<Mutex<Option<ParaId>>>,
+    client_set_aside_for_cidp: Arc<Client>,
+    block_hash: H256,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+where
+    Client: ProvideRuntimeApi<Block>,
+    <Client as ProvideRuntimeApi<Block>>::Api: CollatorAssignmentApi<Block, AccountId, ParaId>,
+{
+    let nimbus_keys = SyncCryptoStore::keys(&*sync_keystore, NIMBUS_KEY_ID).unwrap();
+    let collator_account_id =
```

Comment:
This should not convert to collator_account, it should **always** use the nimbus keys to detect the assignment. Collator accountIds can be very different to their nimbus keys. We should check the assignation with the nimbus Id